SED-1464: Improve reliability of ingestion of large email messages

tajinder opened the pull request.

teji885 committed:
SED-1464: Improve reliability of ingestion of large email messages

teji885 committed:
SED-1464 : stripping prefix and postfix from html using caja library instead of regex
stripping prefix and postfix from style using css library instead of regex

deepwell commented on lib/index.js:
```diff
@@ -39,5 +39,36 @@ function purify(htmlInput, options, cb) {
+ * @param Object options
+ * @cb err, string
+ */
+function revertPurify(htmlInput, options, cb) {
```
We're not really reverting the purify process here, instead we're reverting the namespaced classes and styles. Can you rename this to `revertNamespace`?

deepwell replied:
I didn't see any unit tests for this function, can you add one to ensure this function works? While the other tests you have ensure smaller parts of it function.

deepwell commented on lib/html_parser/index.js:
```diff
@@ -49,5 +49,25 @@ function parse(htmlInput, prefix, postfix, cb) {
+function parsePurified(htmlInput, prefix, postfix, cb) {
+  // strip prefix and postfix from id and class attributes recursively
+  if (typeof prefix !== 'undefined' || typeof postfix !== 'undefined') {
+    async.whilst(function() {
```
I managed to break it by making this while loop run forever. All I did was added the prefix `ugc` to part of the html body that's not in a class name.

deepwell commented on lib/html_parser/namespacer/index.js:
```diff
@@ -89,5 +89,71 @@ function namespace(htmlInput, prefix, postfix, cb) {
+          else if (key === 'class' && value) {
+            hasClass = true;
+            value = value.replace(postfix, '').trim();
+            outputCache.push(' class="');
```
I see `hasClass` is not used anywhere. Ideally if there are no class names left (they all have been removed). We should not add `class=""` to the html output.

deepwell commented on lib/style_parser/namespacer/index.js:
```diff
@@ -35,5 +35,49 @@ function namespace(styles, prefix, postfix, cb) {
+    // strip styles
+    function(prefixedStyles, cb) {
+      if (postfix) {
+        async.whilst(function() {
```
This could cause it to crash as well if the postfix is part of a class name that's not at the beginning.

teji885 committed:
SED-1464 code review comments
added proper checks so the while loop doesn't loop forever
added check so that empty class is removed

deepwell commented on lib/html_parser/index.js:
```diff
@@ -50,22 +50,47 @@ function parse(htmlInput, prefix, postfix, cb) {
+          namespacer.stripNamespace(attributeSanitizedHtml, prefix, postfix, function(err, revertedHtml, replaced, hasClass) {
+            attributeSanitizedHtml = revertedHtml;
+            if (!hasClass) {
+              attributeSanitizedHtml = revertedHtml.replace('class = ""', '')
```
This should be done in the namespacer.

deepwell commented on lib/style_parser/namespacer/prefixer.js:
```diff
@@ -24,15 +24,20 @@ function prepend(styles, prefix, cb) {
   var outputCache = [];
   var selectors = [];
   var rules = [];
+  var isReverted = false;
```
Can you explain what `isReverted` is for and why we need to pass this back?

tajinder replied:
isReverted is used to control the repeating loop , which needs to call prefixer or postfixer until all the prefixes or postfixes are removed.
So it goes to prefixer just one extra time, this time it doesn't find any prefixes and return without reverting anything. And the loop doesn't continue if isReverted is returned false.

teji885 committed:
SED-1464 move check for hasClass to namespacer from index

deepwell commented:
Hi Tajinder, can you please rebase this and the sedna branch on top of master? There have been some updates since this was started.

deepwell committed:
SED-1464 Trim extra whitespace in class names

- Blank spaces were adding in extra prefix classes on namespace

deepwell committed:
SED-1464 Preserve line feeds

deepwell committed:
Add tests for missing edge cases

deepwell commented:
I have found 2 more cases that this code currently fails, I have added them as unit tests in ea74e10c in the branch.

teji885 committed:
SED-1464 allow multiple styles in css
fix to not remove empty class from string
added an extra space after style : for better formatting of clean css

deepwell committed:
SED-1464 remove unneeded test

Commit:
Added failing test for ugc stripping

teji885 committed:
SED-1464 now removing postfix only if its not followed by - because postfix followed by a - is prefix

deepwell commented:
Tajinder is there a unit test we can add to ensure we don't regress on this?

deepwell commented:
Ammar wrote a test here: https://github.com/stage3systems/node-html5-purifier/pull/6/commits/5b11b4bba2d924983a01549962ee285234085f17
